File: a5/ans3.pl
/** 
// Honor Pledge:
//
// I pledge that I have neither given nor 
// received any help on this assignment.
//
// gforsyth 
*/ 

/**
    Name: Gabbi Forsythe
    Date: 11/22/2019
    Class: CS441 
    Purpose: Part 3 of 3 of assignment 5 over prolog. A prolog predicate to perform insertion sort on a list.
*/

sort([H|T], Sorted) :- sort(T, list), !, insertion(H, list, Sorted). /** [H|T] is sorted if, T is sorted and when H is inserted into T (list) it is sorted */
sort(finishedList, finishedList).   /** compares list to make sure they are sorted */

/** adapted from swap algorithm in class */ 
insertion(A, [B|C], [B|D]) :- A > B.    /** compares the value that is passed in (H) to the value in the list other parts of list don't matter */
insertion(A, C, [A|C]) :- insertion(A, C, D).   

File: a6/Ans3.ml
(**
// Honor Pledge:
//
// I pledge that I have neither given nor 
// received any help on this assignment.
//
// gforsyth 
*)

(**
    Name: Gabbi Forsythe
    Date: 12/13/2019
    Class: CS441 
    Purpose: Part 3 of 3 of assignment 6 over ML. A SML program that performs insertion sort on a list. 
*)

fun insert x []= [x]     (* If the list is empty *)
    | insert x (y::tailY)=  if x < y 
                            then x :: y :: tailY        (* If x is less than the head of the list, put x at the head of the list *)
                            else y :: (insert x tailY);       (* If x is greater than the front of the list keep calling the insert funciton until x can be put in the proper place*)


fun sort([]) = []       (* Base Case: Do the sorting until the original list is empty *)
    | sort (x::tailX) = insert x (sort tailX); (* Take the values of the list off until all the values have been removed from the list, then begin inserting them *)